android/shared/src/main/kotlin/kmp/android/shared/navigation/Destination.kt

```diff
@@ -0,0 +1,149 @@
+package kmp.android.shared.navigation
+
+import android.net.Uri
+import androidx.compose.foundation.layout.ColumnScope
+import androidx.compose.runtime.Composable
+import androidx.navigation.NamedNavArgument
+import androidx.navigation.NavBackStackEntry
+import androidx.navigation.NavDeepLink
+import androidx.navigation.NavGraphBuilder
+import androidx.navigation.compose.composable
+import androidx.navigation.compose.dialog
+import com.google.accompanist.navigation.material.ExperimentalMaterialNavigationApi
+import com.google.accompanist.navigation.material.bottomSheet
+
+/**
+ * Base class for navigation graph root class.
+ */
+interface RootDestination {
+    val route: String
+}
+
+/**
+ * Base class for any destination the app uses.
+ */
+abstract class Destination(parent: RootDestination?) {
```

Comment:
1. I thought just using as many `RootDestination`s as needed should be okay, but maybe I should add the path stacking from parents, since now only the Destinations get path prefixes from one parent. 
Why would a middle child have arguments? I can't imagine a use case, since I imaged all the `RootDestination`s would be roots of nav graphs containing `Destination`s and not corresponding to actual screens.
2. When using it when you want to access the arguments you can see them as parameters of the Args class
`val args = UsersDestination.Detail.Args(navBackStackEntry.arguments)` and then `args.userId`. But yeah, it's not that easy to access as one would wish.

Reply:
Maybe I misunderstood, did you mean having Root something like Users, then UserList as a Destination and then as a child of the UserList for example UserDetail? So that you would have root > child > child? This use case we used as twi children of the same route, since this hierarchy only show the navigation trees, not the whole hierarchy of screens (which can be accessed from which). I think it is same with the native android navigation, you don't have children with children of their own. Does it make sense? Or am I overlooking a use case when we would need this?

Reply:
>did you mean having Root something like Users, then UserList as a Destination and then as a child of the UserList for example UserDetail?  

yes, that's what I had in mind. And yeah, I am just looking for edge cases that might never be relevant 😅 One example from Heureka could be a root "Home" destination with "ProductDetail" child which has `productId` as parameter and then "OfferDetail" destination which is child of "ProductDetail" and with the `productId` has one more `offerId` argument. So we have `Home > ProductDetail(productId) > OfferDetail(productId, offerId)` relationship.   
I agree that this can easily be solved by having all destinations be child of the "Home". Just that the existing root-child relationship pushes me into thinking about it as a hierarchy but not actually enabling it. Again, I am not saying that this is wrong, just that **subjectively** the naming and API of this implies hierarchy and not providing it.   

And I am not sure what you mean by "native android navigation" but I think you can have several nested nav graphs and the routes are just strings so it can "support" repeated nesting.

Reply:
By "native android navigation" I meant the AndroidX navigation (I consider it native, since its by google and not a third party library) 🙂 Maybe I am mistaken but I think you cannot have there a screen node (composable) as a parent of other screen nodes, the parents are not tied to screens.

Reply:
I was confused because we are still using the native navigation so I thought you meant something else.   
And yes, there is no concept of parent and child relationship in native navigation at all. You can only nest navigation graphs.  
So how would you implement the Heureka example? I've added one more destination:
- Home: `home`
- Product detail: `home/product/123`
- Offer detail: `home/product/123/offer/456`
- Shop exit: `home/product/123/exit/789`

Reply:
I guess I would simply have a root and inside all these destinations:
- Home without arguments
- Product Detail with productId argument
- Offer Detail with productId and offerId arguments
- Shop Exit with productId and exitId arguments
And if you wanted to go from Product Detail to Offer Detail, you would pass in the productId that has been already passed to the Product Detail screen and add the offerId

I get your point, but I'm not sure we would benefit from hierarchical destinations. In case you suddenly needed to get to Offer Detail from some other place, the hierarchy would be in the way. Also there would be two sources of truth: one being the hierarchy of destinations and the other being the real backstack, so I really don't know about that.